src/main/java/org/truffleruby/language/SafepointManager.java

```diff
@@ -270,4 +292,18 @@ private void interruptOtherThreads() {
         }
     }
 
+    public void shutdown() {
+        if (!runningThreads.isEmpty()) {
+            Log.LOGGER.warning("still threads registered with safepoint manager at shutdown:\n" + context.getThreadManager().getThreadDebugInfo() + getSafepointDebugInfo());
+        }
+    }
+
+    public String getSafepointDebugInfo() {
+        final Thread[] threads = new Thread[Thread.activeCount() + 1024];
```

Comment:
1024 seems excessive, I think it would better to retry if the return value of enumerate is lower than the activeCount().
https://docs.oracle.com/javase/7/docs/api/java/lang/Thread.html#enumerate(java.lang.Thread[])

Reply:
It's a race condition. You could keep retrying forever.